File: src/main/java/com/my/admin/app/commerce/service/CashService.java
package com.my.admin.app.commerce.service;

import com.black.food.manager.dao.CashDao;
import com.black.food.manager.dao.TurnDao;
import com.black.food.manager.dao.UserDao;
import com.black.food.manager.exception.ApplicationException;
import com.black.food.manager.helper.DateHelper;
import com.black.food.manager.model.api.request.cash.CreateCashRequest;
import com.black.food.manager.model.api.request.cash.UpdateCashRequest;
import com.black.food.manager.model.api.response.CashResponse;
import com.black.food.manager.model.api.response.CashStatusResponse;
import com.black.food.manager.model.entity.Cash;
import com.black.food.manager.model.entity.CashStatus;
import com.black.food.manager.model.entity.Turn;
import com.black.food.manager.model.entity.User;
import com.black.food.manager.model.transformer.CashTransformer;
import com.black.food.manager.service.validator.cash.CashFluxValidator;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import java.time.ZonedDateTime;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class CashService {

    private CashDao cashDao;
    private UserDao userDao;
    private TurnDao turnDao;
    private DateHelper dateHelper;

    private CashFluxValidator cashFluxValidator;
    private CashTransformer cashTransformer;

    private static final Logger LOGGER = LoggerFactory.getLogger(CashService.class);

    public CashResponse createCash(final CreateCashRequest createCashRequest) {

        cashFluxValidator.validateRequestForCreate(createCashRequest);

        User user = findUserById(createCashRequest.getUserId());
        Turn turn = findTurnById(createCashRequest.getTurnId());

        Cash cash = new Cash();

        cash.setUser(user);
        cash.setTurn(turn);
        cash.setStartDate(getNowZonedDateTime());
        cash.setStartMoney(createCashRequest.getStartMoney());
        cash.setCashStatus(findCashStatus(createCashRequest.getStatusId()));

        return cashTransformer.apply(save(cash));
    }

    public CashResponse update(final String id, final UpdateCashRequest updateCashRequest) {
        cashFluxValidator.validateRequestForUpdate(updateCashRequest);

        Cash cashReturn;
        Cash cashToUpdate = findById(Long.valueOf(id));
        try {
            var cashStatus = updateCashRequest.getStatusId() == null
                    ? cashToUpdate.getCashStatus()
                    : findCashStatusById(Long.valueOf(updateCashRequest.getStatusId()));

            cashToUpdate.setStartMoney(updateCashRequest.getStartMoney());
            cashToUpdate.setStartDate(updateCashRequest.getStartDate());
            cashToUpdate.setCashStatus(cashStatus);

            cashReturn = save(cashToUpdate);
        } catch (Exception exception) {
            var error = String.format("Error while updating cash whit id %s", id);
            throw throwConflictException(exception.getLocalizedMessage(), error);
        }

        return cashTransformer.apply(cashReturn);
    }

    public Set<CashResponse> findAll() {
        return cashDao.findAll()
                .stream()
                .map(cash -> cashTransformer.apply(cash))
                .collect(Collectors.toSet());
    }

    public Cash findById(final Long id) {
        return cashDao.findById(id)
                .orElseThrow(() -> throwNotFoundException(id, "id"));
    }

    public Set<CashResponse> findByStatus(final Long cashStatusId) {
        var activeCashes = cashDao.findByStatus(cashStatusId);

        activeCashes
                .stream()
                .findFirst()
                .orElseThrow(() -> throwNotFoundException(cashStatusId, "status id"));

        return activeCashes
                .stream()
                .map(cash -> cashTransformer.apply(cash))
                .collect(Collectors.toSet());
    }

    public CashResponse findCashResponseById(final Long id) {
        return cashTransformer.apply(findById(id));
    }

    public CashResponse deleteCash(final String id) {

        Cash cash = findById(Long.valueOf(id));

        try {
            cashDao.delete(cash);
        } catch (Exception exception) {
            var messageError = exception.getLocalizedMessage();
            var error = String.format("Error while deleting cash with id %d.", cash.getId());
            throw throwConflictException(messageError, error);
        }

        return cashTransformer.apply(cash);
    }

    public Set<CashStatusResponse> findAllCashStatus() {
        return cashDao.findAllCashStatus()
                .stream()
                .map(cashStatus -> cashTransformer.transformCashStatusToCashStatusResponse(cashStatus))
                .collect(Collectors.toSet());
    }

    private Cash save(final Cash cash) {
        Cash createdCash;
        try {
            createdCash = cashDao.save(cash);
        } catch (Exception exception) {
            var error = "Error while saving cash.";
            throw throwConflictException(exception.getLocalizedMessage(), error);
        }
        return createdCash;
    }

    private CashStatus findCashStatus(final String cashStatusId) {
        return findCashStatusById(Long.valueOf(cashStatusId));
    }

    private CashStatus findCashStatusById(final Long cashStatusId) {
        return cashDao.findCashStatusById(cashStatusId)
                .orElseThrow(() -> throwNotFoundExceptionToCashStatus(cashStatusId, "id"));
    }

    private ZonedDateTime getNowZonedDateTime() {
        return dateHelper.getNowZonedDateTime();
    }

    public User findUserById(final String userId) {
        return userDao.findById(Long.valueOf(userId)).orElseThrow();
    }

    public Turn findTurnById(final String turnId) {
        return turnDao.findById(Long.valueOf(turnId)).orElseThrow();
    }

    private RuntimeException throwNotFoundException(final Long id, final String param) {
        var message = String.format("Cash with %s %s not found", param, id);
        LOGGER.error(message);
        throw new ApplicationException(message, HttpStatus.NOT_FOUND);
    }

    private RuntimeException throwNotFoundExceptionToCashStatus(final Long id, final String param) {
        var message = String.format("Cash status with %s %s not found", param, id);
        LOGGER.error(message);
        throw new ApplicationException(message, HttpStatus.NOT_FOUND);
    }

    private RuntimeException throwConflictException(final String message, final String error) {
        var completeMessage = String.format("Error : %s", message);
        LOGGER.error(error);
        LOGGER.error(completeMessage);
        throw new ApplicationException(completeMessage, HttpStatus.CONFLICT);
    }

    @Autowired
    public void setCashDao(final CashDao cashDao) {
        this.cashDao = cashDao;
    }

    @Autowired
    public void setUserDao(final UserDao userDao) {
        this.userDao = userDao;
    }

    @Autowired
    public void setTurnDao(final TurnDao turnDao) {
        this.turnDao = turnDao;
    }

    @Autowired
    public void setDateHelper(final DateHelper dateHelper) {
        this.dateHelper = dateHelper;
    }

    @Autowired
    public void setCashFluxValidator(final CashFluxValidator cashFluxValidator) {
        this.cashFluxValidator = cashFluxValidator;
    }

    @Autowired
    public void setCashTransformer(final CashTransformer cashTransformer) {
        this.cashTransformer = cashTransformer;
    }
}

File: src/main/kotlin/com/bk/reactive/app/commerce/admin/my/service/CashService.kt
package com.bk.reactive.app.commerce.admin.my.service

import com.bk.reactive.app.commerce.admin.my.dao.CashDAO
import com.bk.reactive.app.commerce.admin.my.dao.TurnDAO
import com.bk.reactive.app.commerce.admin.my.dao.UserDAO
import com.bk.reactive.app.commerce.admin.my.exception.api.SaveCashException
import com.bk.reactive.app.commerce.admin.my.exception.api.SaveSaleException
import com.bk.reactive.app.commerce.admin.my.helper.DateHelper
import com.bk.reactive.app.commerce.admin.my.model.api.request.cash.CreateCashRequest
import com.bk.reactive.app.commerce.admin.my.model.api.request.cash.FindCashRequest
import com.bk.reactive.app.commerce.admin.my.model.api.request.cash.UpdateCashRequest
import com.bk.reactive.app.commerce.admin.my.model.api.response.CashResponse
import com.bk.reactive.app.commerce.admin.my.model.api.response.CashStatusResponse
import com.bk.reactive.app.commerce.admin.my.model.api.response.SaleResponse
import com.bk.reactive.app.commerce.admin.my.model.document.Cash
import com.bk.reactive.app.commerce.admin.my.model.document.CashStatus
import com.bk.reactive.app.commerce.admin.my.model.document.Turn
import com.bk.reactive.app.commerce.admin.my.model.document.User
import com.bk.reactive.app.commerce.admin.my.model.mapper.CashMapper
import com.bk.reactive.app.commerce.admin.my.service.validator.cash.CashFluxValidator
import mu.KotlinLogging
import org.springframework.http.HttpStatus
import org.springframework.stereotype.Service
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono
import reactor.kotlin.core.util.function.component1
import reactor.kotlin.core.util.function.component2
import reactor.kotlin.core.util.function.component3
import java.time.ZonedDateTime

@Service
class CashService(
    private val cashDao: CashDAO,
    private val userDao: UserDAO,
    private val turnDao: TurnDAO,
    private val dateHelper: DateHelper,
    private val cashFluxValidator: CashFluxValidator,
    private val exceptionService: ExceptionService
) {
    private val logger = KotlinLogging.logger { }

    fun createCash(createCashRequest: CreateCashRequest): Mono<CashResponse> {
        return cashFluxValidator.validateRequestForCreate(createCashRequest)
            .flatMap {
                Mono.zipDelayError(
                    findUserById(createCashRequest.userId),
                    findTurnById(createCashRequest.turnId),
                    findCashStatusById(createCashRequest.statusId)
                )
            }.flatMap { (user, turn, cashStatus) ->
                save(
                    Cash(
                        user = user,
                        turn = turn,
                        startDate = getNowZonedDateTime(),
                        startMoney = createCashRequest.startMoney,
                        cashStatus = cashStatus
                    )
                )
            }.map { cash ->
                CashMapper.fromCashToCashResponse(cash)
            }.doOnNext { cashSaved ->
                logger.info { "Cash with id ${cashSaved.id} saved" }
            }.onErrorSaveCashResume()
    }

    fun update(id: String, updateCashRequest: UpdateCashRequest): Mono<CashResponse> {
        return cashFluxValidator.validateRequestForUpdate(updateCashRequest)
            .flatMap { findById(id) }
            .flatMap { cashToUpdate ->
                if (updateCashRequest.statusId == cashToUpdate.cashStatus.id)
                    Mono.zip(
                        Mono.just(cashToUpdate),
                        Mono.just(cashToUpdate.cashStatus)
                    )
                else Mono.zip(
                    Mono.just(cashToUpdate),
                    findCashStatusById(updateCashRequest.statusId)
                )
            }.flatMap { (cashToUpdate, cashStatus) ->
                save(
                    Cash(
                        id = cashToUpdate.id,
                        turn = cashToUpdate.turn,
                        user = cashToUpdate.user,
                        cashStatus = cashStatus,
                        startMoney = updateCashRequest.startMoney,
                        startDate = updateCashRequest.startDate,
                        partialBalance = cashToUpdate.partialBalance
                    )
                )
            }.map { savedCash ->
                CashMapper.fromCashToCashResponse(savedCash)
            }.onErrorResume {
                val error = "Error while updating cash whit id $id"
                logger.error { error }
                Mono.error { exceptionService.throwConflictException(it.localizedMessage, error) }
            }
    }

    fun updatePartialBalance(id: String, cash: Cash): Mono<Cash> {
        return findById(id)
            .flatMap {
                save(
                    Cash(
                        id = cash.id,
                        turn = cash.turn,
                        user = cash.user,
                        cashStatus = cash.cashStatus,
                        startMoney = cash.startMoney,
                        startDate = cash.startDate,
                        partialBalance = cash.partialBalance
                    )
                )
            }
    }

    fun cashFilter(
        findCashRequest: FindCashRequest
    ): Flux<CashResponse> {
        return cashDao
            .findAll()
            .filterCashes(findCashRequest)
            .map { CashMapper.fromCashToCashResponse(it) }
            .onErrorResume {
                val error = "Error while sending cashes by filters"
                logger.error { error }
                Mono.error { exceptionService.throwNotFoundException(it.localizedMessage, error) }
            }
    }

    fun findCashResponseById(id: String): Mono<CashResponse> {
        return findById(id)
            .map { CashMapper.fromCashToCashResponse(it) }
    }

    fun deleteCash(id: String): Mono<CashResponse> {
        return findById(id)
            .flatMap { cash ->
                cashDao.delete(cash).flatMap { Mono.just(cash) }
            }.map { cashDeleted ->
                CashMapper.fromCashToCashResponse(cashDeleted)
            }.onErrorResume { error ->
                val message = "Error while deleting cash with id ${id}."
                logger.error { message }
                Mono.error { exceptionService.throwConflictException(error.localizedMessage, message) }
            }.doOnNext {
                logger.info { "Cash with id $id deleted" }
            }
    }

    fun findById(id: String): Mono<Cash> {
        return cashDao.findById(id)
            .switchIfEmpty(
                Mono.error { exceptionService.throwNotFoundException(id, "cashId") })
            .onErrorResume {
                logger.error { it.localizedMessage }
                Mono.error { it }
            }
    }

    fun findAllCashStatus(): Flux<CashStatusResponse> {
        return cashDao.findAllCashStatus()
            .map { cashStatus -> CashMapper.fromCashStatusToCashStatusResponse(cashStatus) }
    }

    private fun save(cash: Cash): Mono<Cash> {
        return cashDao.save(cash)
            .doOnError {
                logger.error { "An error occurred while save a cash : $cash" }
            }
    }

    //ATBBfMqtNUQEcKeaLSSPnhBxWB3kF3EE0B53 bitbuket token
    private fun findCashStatusById(cashStatusId: String): Mono<CashStatus> {
        return cashDao.findCashStatusById(cashStatusId)
            .switchIfEmpty(
                Mono.error { exceptionService.throwNotFoundException(cashStatusId, "cashStatusId") })
            .onErrorResume {
                logger.error { it.localizedMessage }
                Mono.error { it }
            }
    }

    private fun findUserById(userId: String): Mono<User> {
        return userDao.findById(userId)
            .switchIfEmpty(
                Mono.error { exceptionService.throwNotFoundException(userId, "userId") })
            .onErrorResume {
                logger.error { it.localizedMessage }
                Mono.error { it }
            }
    }

    private fun findTurnById(turnId: String): Mono<Turn> {
        return turnDao.findById(turnId)
            .switchIfEmpty(
                Mono.error { exceptionService.throwNotFoundException(turnId, "turnId") })
            .onErrorResume {
                logger.error { it.localizedMessage }
                Mono.error { it }
            }
    }

    private fun Flux<Cash>.filterCashes(
        findCashRequest: FindCashRequest
    ) = this.filter { cash -> findCashRequest.id == null || cash.id == findCashRequest.id }
        .filter { cash -> findCashRequest.cashStatusId == null || cash.cashStatus.id == findCashRequest.cashStatusId }
        .filter { cash -> findCashRequest.turnId == null || cash.turn.id == findCashRequest.turnId }
        .filter { cash -> findCashRequest.userId == null || cash.user.id == findCashRequest.userId }
        .filter { cash -> findCashRequest.dateFrom == null || cash.startDate >= ZonedDateTime.parse(findCashRequest.dateFrom) }
        .filter { cash -> findCashRequest.dateTo == null || cash.startDate <= ZonedDateTime.parse(findCashRequest.dateTo) }

    private fun Mono<CashResponse>.onErrorSaveCashResume(): Mono<CashResponse> =
        this.onErrorResume {
            logger.error { it.printStackTrace() }
            Mono.error {
                SaveCashException(
                    code = HttpStatus.CONFLICT,
                    message = it.localizedMessage,
                    suppressedSaleException = it.suppressed
                )
            }
        }

    private fun getNowZonedDateTime(): ZonedDateTime = dateHelper.getNowZonedDateTime()
}